### DISCUSSION TITLE
Reducing single-letter command-line arguments

### DISCUSSION THREAD

**Discussion by vpax** [zeek/zeek#2061] [Ideas]
Right now, `zeek -h` documents 35 different single-letter command-line arguments.  At least some of these came about inadvertently because `Options.cc` uses `getopt_long()` in a way that requires every long option to have a short command associated with it.

I'd like to significantly reduce the set of single-letter options.  I'm thinking we keep the following 14:

> -b|--bare-mode
-d|--debug-script  (I'm not sure whether this even still works)
-e|--exec
-f|--filter
-h|--help
-i|--iface
-r|--readfile
-t|--tracefile
-v|--version
-w|--writefile  (is this used much anymore?)
-B|--debug
-C|--no-checksums
-N|--print-plugins
-O|--optimize

and change the following 21 to be long-option-only (with the single-letter versions deprecated for Zeek 5):

> -a|--parse-only
-c|--capture-unprocessed
-p|--prefix   (I believe this isn't used much these days)
-s|--rulefile
-u|--usage-issues
-D|--deterministic
-E|--event-trace
-F|--force-dns
-G|--load-seeds
-H|--save-seeds
-I|--print-id
-0|--optimize-files=<pat>
-o|--optimize-funcs=<pat>
-P|--prime-dns
-Q|--time    (didn't even know this existed!  Seems busted on MacOS)
-S|--debug-rules
-T|--re-level   (still relevant?)
-U|--status-file   (what the heck is this? can't find any documentation on it)
-W|--watchdog
-X|--zeekygen
-j|--jobs

**Comment by rsmmr** [comment]
Is the motivation to reuse them for other purposes?

Generally seems fine with me, but we should indeed go through the deprecation cycle. Also, I'd keep these as well:

```
-s|--rulefile (pretty commonly used)
-D|--deterministic (used regularly in tests)
-G|--load-seeds (used regularly in tests)
-j|--jobs (introduced by new supervisor, will eventually have quite standard `-j N` meaning for parallelization)
```

> `-U|--status-file (what the heck is this? can't find any documentation on it)````

This communicates process status to ZeekControl. If we remove this, ZeekControl needs to be adapted.

**Reply by vpax** [reply]
> Is the motivation to reuse them for other purposes?

That, when appropriate, and to simplify the `--help` output and have it more directly point up which options are used commonly vs. those that are more niche.

> Also, I'd keep these as well:

Okay by me.

> This communicates process status to ZeekControl. If we remove this, ZeekControl needs to be adapted.

Is it likely that it's only used by ZeekControl?  If so, then easy enough to fix that usage.

**Reply by rsmmr** [reply]
> That, when appropriate, and to simplify the `--help` output and have it more directly point up which options are used commonly vs. those that are more niche.

The latter could also be achieved by reorganizing the `--help`  output into different sections ("common options", "advanced options"). And/or: hide the advanced options by default and introduce something like ``--help-all``. That would allow us to be less invasive by letting more options stay the way they are (I'm not really concerned either way, just pointing it out).


> > This communicates process status to ZeekControl. If we remove this, ZeekControl needs to be adapted.

> Is it likely that it's only used by ZeekControl? If so, then easy enough to fix that usage.

Yes (my comment was meant primarily as "let's remember this when we do the change").
